pubsub/google/cloud/pubsub_v1/subscriber/_protocol/leaser.py

```diff
@@ -0,0 +1,186 @@
+                self._bytes += item.byte_size
+            else:
+                _LOGGER.debug(
+                    'Message %s is already lease managed', item.ack_id)
+
+    def remove(self, items):
+        """Remove messages from lease management."""
+        # Remove the ack ID from lease management, and decrement the
+        # byte counter.
+        for item in items:
+            if self._leased_messages.pop(item.ack_id, None) is not None:
+                self._bytes -= item.byte_size
+            else:
+                _LOGGER.debug('Item %s was not managed.', item.ack_id)
+
+        if self._bytes < 0:
+            _LOGGER.debug(
+                'Bytes was unexpectedly negative: %d', self._bytes)
+            self._bytes = 0
+
+    def maintain_leases(self):
+        """Maintain all of the leases being managed by the subscriber.
+
+        This method modifies the ack deadline for all of the managed
+        ack IDs, then waits for most of that time (but with jitter), and
+        repeats.
+        """
+        while self._subscriber.is_active and not self._stop_event.is_set():
+            # Determine the appropriate duration for the lease. This is
+            # based off of how long previous messages have taken to ack, with
+            # a sensible default and within the ranges allowed by Pub/Sub.
+            p99 = self._subscriber.ack_histogram.percentile(99)
+            _LOGGER.debug('The current p99 value is %d seconds.', p99)
+
+            # Make a copy of the leased messages. This is needed because it's
+            # possible for another thread to modify the dictionary while
+            # we're iterating over it.
+            leased_messages = copy.copy(self._leased_messages)
+
+            # Drop any leases that are well beyond max lease time. This
+            # ensures that in the event of a badly behaving actor, we can
+            # drop messages and allow Pub/Sub to resend them.
+            cutoff = (
+                time.time() -
+                self._subscriber.flow_control.max_lease_duration)
+            to_drop = [
+                requests.DropRequest(ack_id, item.size)
+                for ack_id, item
+                in six.iteritems(leased_messages)
+                if item.added_time < cutoff]
+
+            if to_drop:
+                _LOGGER.warning(
+                    'Dropping %s items because they were leased too long.',
+                    len(to_drop))
+                self._subscriber.drop(to_drop)
+
+            # Remove dropped items from our copy of the leased messages (they
+            # have already been removed from the real one by self.drop).
+            for item in to_drop:
+                leased_messages.pop(item.ack_id)
+
+            # Create a streaming pull request.
+            # We do not actually call `modify_ack_deadline` over and over
+            # because it is more efficient to make a single request.
+            ack_ids = list(leased_messages.keys())
```

Comment:
as best I can tell this doesn't really need to be a list. Does this have something to do with back-compat?